### PR TITLE
kd tree data structure implementation

### PR DESCRIPTION
**Pull request by Ramy-Badr-Ahmed**
### Describe your change:

This pull request adds an implementation of the KD-Tree data structure along with related functionalities, such as nearest neighbor searches.

**Definition**: The KD-Tree is a space-partitioning data structure used for organizing points in a k-dimensional space. (URL: [KD-Tree](https://en.wikipedia.org/wiki/K-d_tree))

**Implementation Details**:

- `build_kdtree.py`: Contains the function to build a KD-Tree from a set of points.
- `nearest_neighbor_search.py`: Implements the nearest neighbor search algorithm using the built KD-Tree.
- `kd_node.py`: Defines the KDNode class representing nodes in the KD-Tree.

**Example Usage**:
- `hypercube_points.py`: Provides a function to generate random points uniformly distributed within an n-dimensional hypercube.
- `example_usage.py`: Demonstrates how to use the KD-Tree implementation with the generated random hypercube points and querying for nearest neighbors.

### CODE REVIEW

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
done

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
Added pytests, docstrings, & typehints

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
added pytests, changed assertion style

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
eliminated earlier build errors

**Comment by Ramy-Badr-Ahmed** [comment]
Fixed pre-commit.ci - pr & ruff / ruff (pull_request) checks

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
All checks have passed

**Review by imSanko — CHANGES_REQUESTED** [review]
Sound Good to me, [at]Claus will let you know better.

**Comment by Ramy-Badr-Ahmed** [comment]
@cclauss 
Thanks for your review :)

**Comment by Ramy-Badr-Ahmed** [comment]
> Sound Good to me, [at]Claus will let you know better.

Thanks! The "tests are failing" label is till assigned.

File: data_structures/kd_tree/build_kdtree.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    """
+    Builds a KD-Tree from a list of points.
+
+    Args:
+        points (list[list[float]]): The list of points to build the KD-Tree from.
+        depth (int): The current depth in the tree
+                     (used to determine axis for splitting).
+
+    Returns:
+        KDNode | None: The root node of the KD-Tree,
```
Comment (cclauss) [inline data_structures/kd_tree/build_kdtree.py:14]:
```suggestion
        The root node of the KD-Tree,
```

File: data_structures/kd_tree/example/hypercube_points.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+def hypercube_points(
+    num_points: int, hypercube_size: float, num_dimensions: int
+) -> np.ndarray:
+    """
+    Generates random points uniformly distributed within an n-dimensional hypercube.
+
+    Args:
+        num_points (int): Number of points to generate.
+        hypercube_size (float): Size of the hypercube.
+        num_dimensions (int): Number of dimensions of the hypercube.
```
Comment (cclauss) [inline data_structures/kd_tree/example/hypercube_points.py:13]:
```suggestion
        num_points: Number of points to generate.
        hypercube_size: Size of the hypercube.
        num_dimensions: Number of dimensions of the hypercube.
```

File: data_structures/kd_tree/example/hypercube_points.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    Returns:
+        np.ndarray: An array of shape (num_points, num_dimensions)
```
Comment (cclauss) [inline data_structures/kd_tree/example/hypercube_points.py:16]:
```suggestion
        An array of shape (num_points, num_dimensions)
```

File: data_structures/kd_tree/tests/test_kdtree.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+def test_build_kdtree():
+    """
+    Test that KD-Tree is built correctly.
+    """
+    num_points = 10
+    cube_size = 10.0
+    num_dimensions = 2
+    points = hypercube_points(num_points, cube_size, num_dimensions)
+    kdtree = build_kdtree(points.tolist())
```
Comment (cclauss) [inline data_structures/kd_tree/tests/test_kdtree.py:17]: Please add `build_kdtree()` tests where:
* [ ] `points = []`
* [ ] `depth = 2`
* [ ] `depth = -2`

Comment (Ramy-Badr-Ahmed) [reply data_structures/kd_tree/tests/test_kdtree.py]: done

File: data_structures/kd_tree/tests/test_kdtree.py [hunk 5]
```diff
@@ -1,29 +1,48 @@
...
+        assert kdtree is None, f"Expected None for empty points list, got {kdtree}"
+    else:
+        # Check if root node is not None
+        assert kdtree is not None, "Expected a KDNode, got None"
 
-    # Check if root has correct dimensions
-    assert len(kdtree.point) == num_dimensions
+        # Check if root has correct dimensions
+        assert len(kdtree.point) == num_dimensions, \
+            f"Expected point dimension {num_dimensions}, got {len(kdtree.point)}"
```
Comment (cclauss) [inline data_structures/kd_tree/tests/test_kdtree.py]: PEP8: Avoid backslashes in Python code.
```suggestion
        assert len(kdtree.point) == num_dimensions, (
            f"Expected point dimension {num_dimensions}, got {len(kdtree.point)}"
        )
```

File: data_structures/kd_tree/tests/test_kdtree.py [hunk 6]
```diff
@@ -38,12 +40,14 @@ def test_build_kdtree(num_points, cube_size, num_dimensions, depth, expected_res
         assert kdtree is not None, "Expected a KDNode, got None"
 
         # Check if root has correct dimensions
-        assert len(kdtree.point) == num_dimensions, \
-            f"Expected point dimension {num_dimensions}, got {len(kdtree.point)}"
+        assert (
```
Comment (Ramy-Badr-Ahmed) [inline data_structures/kd_tree/tests/test_kdtree.py:43]: removed backslashes: PEP8

File: data_structures/kd_tree/build_kdtree.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+from data_structures.kd_tree.kd_node import KDNode
+
+
+def build_kdtree(points: list[list[float]], depth: int = 0) -> KDNode | None:
+    """
+    Builds a KD-Tree from a list of points.
+
+    Args:
+        points (list[list[float]]): The list of points to build the KD-Tree from.
+        depth (int): The current depth in the tree
```
Comment (cclauss) [inline data_structures/kd_tree/build_kdtree.py:10]:
```suggestion
        points: The list of points to build the KD-Tree from.
        depth: The current depth in the tree
```

File: data_structures/kd_tree/kd_node.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+from typing import Optional
+
+
```
Comment (cclauss) [inline data_structures/kd_tree/kd_node.py:3]:
```suggestion
```

File: data_structures/kd_tree/kd_node.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        point (list[float]): The point stored in this node.
+        left (Optional[KDNode]): The left child node.
+        right (Optional[KDNode]): The right child node.
+    """
+
+    def __init__(
+        self,
+        point: list[float],
+        left: Optional["KDNode"] = None,
+        right: Optional["KDNode"] = None,
```
Comment (cclauss) [inline data_structures/kd_tree/kd_node.py:18]:
```suggestion
        point: The point stored in this node.
        left: The left child node.
        right: The right child node.
    """

    def __init__(
        self,
        point: list[float],
        left: KDNode | None = None,
        right: KDNode | None  = None,
```

File: data_structures/kd_tree/nearest_neighbour_search.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    nearest_dist: float = float("inf")
+    nodes_visited: int = 0
+
+    def search(node: KDNode | None, depth: int = 0) -> None:
+        """
+        Recursively searches for the nearest neighbor in the KD-Tree.
+
+        Args:
+            node (KDNode | None): The current node in the KD-Tree.
+            depth (int): The current depth in the KD-Tree.
```
Comment (cclauss) [inline data_structures/kd_tree/nearest_neighbour_search.py:32]:
```suggestion
            node: The current node in the KD-Tree.
            depth: The current depth in the KD-Tree.
```

**Comment by cclauss** [comment]
I made some changes... Let's see if it passes.

**Comment by Ramy-Badr-Ahmed** [comment]
> I made some changes... Let's see if it passes.

Works now :-)

**Review by cclauss — APPROVED** [review]
Nice...  Nodes are usually great candidates for `dataclass`.

Thanks for doing this.

**Pull request merged**